README.org
* Cask Package Toolset

[[https://travis-ci.org/AdrieanKhisbe/cask-package-toolset.el?branch=master][file:https://travis-ci.org/AdrieanKhisbe/cask-package-toolset.el.svg?branch=master]]
[[https://coveralls.io/r/AdrieanKhisbe/cask-package-toolset.el][file:https://coveralls.io/repos/AdrieanKhisbe/cask-package-toolset.el/badge.svg]]
[[http://melpa.org/#/cask-package-toolset][file:http://melpa.org/packages/cask-package-toolset-badge.svg]]
[[http://stable.melpa.org/#/cask-package-toolset][file:http://stable.melpa.org/packages/cask-package-toolset-badge.svg]]
[[https://github.com/AdrieanKhisbe/cask-package-toolset.el/tags][file:https://img.shields.io/github/tag/AdrieanKhisbe/cask-package-toolset.el.svg]]
[[http://www.gnu.org/licenses/gpl-3.0.html][file:http://img.shields.io/:license-gpl3-blue.svg]]


/Continuous Integration for EmacsLisp without pain!/

** About

The goal of *Cask Package Toolset* is to promote the use of integration and tests to
have even better emacs library, where it's more easy to contribute and build new features! :)
All that without fearing regressions, and unwanted breaking changes.

It does that by *helping you to start or tool up your emacs package project* with
best continuous integration practices and emacs tools.
Among which [[travis-ci.org][travis]], [[https://github.com/rejeep/evm][evm]], [[https://github.com/rejeep/ert-runner.el][ert-runner]], [[https://github.com/ecukes/ecukes][ecukes]], [[https://github.com/cask/cask][Cask]], [[https://github.com/sviridov/undercover.el][undercover]], [[https://coveralls.io/][coveralls]].

This is inspired from the [[http://rejeep.github.io/emacs/cask/ert/ert-runner/ert-async/ecukes/testing/travis/2014/01/09/various-testing-tools-in-emacs.html][following blog post]] of [[https://github.com/rejeep/][@Rejeep]] who had brought to us [[https://github.com/cask/cask][Cask]],
[[https://github.com/rejeep/ert-runner.el][Ert-runner]] and [[https://github.com/rejeep/evm][Emacs Version Management]] (and many more :)).

** Installation

Just add this has dependency to your =Cask= file

: (depends-on "cask-package-toolset")

** Usage
+ View usage information:
  : $ cask exec package-toolset help
+ View package status:
  : $ cask exec package-toolset status
+ Install all the tools
  : $ cask exec package-toolset install
+ Print the melpa recipe for your repository
  : $ cask exec package-toolset melpa-recipe
+ Print the badges for the =README=
  : $ cask exec package-toolset badge
+ Setup unit test with [[https://www.gnu.org/software/emacs/manual/html_node/ert/][ert]] and [[https://github.com/rejeep/ert-runner.el][ert-runner]]:
  : $ cask exec package-toolset setup-ert
+ Get the code to setup coverage with [[https://github.com/sviridov/undercover.el][undercover]]
  : $ cask exec package-toolset setup-undercover
+ Update travis config if needed with update-travis
  : $ cask exec package-toolset update-travis

*** Options
+ In case your =origin= remote is not targeting your github repository, you can specify
  it with the =-r <remote>= or =--remote <remote>= option
+ You could otherwise directly specify your repositoryname with the =-g= or =--github-repo=
  option. (don't forget your yoursername or put a stub one ;) =-g toto/secret.el=)
+ For the badges creation, you can specify the syntax you want with =-s <syntax>= or
  =--syntax <syntax>=. Legit values are =orgmode=, =markdown=, =rst= and =html=, =markown= being the default

** Features
# TODO: add status
# §IDEA: badge for CI level?
*** Skeleton File List

Here is the list of the scaffold file to set up continuous integration for your emacs package.

+ The =setup=, =install= command is to be run first, and will setup the following files
  - [[file:templates/Makefile][Makefile]] : The =Makefile=, and it's =test= targets
  - [[file:templates/.gitignore][.gitgnore]] : A classic =gitignore= designed for a /casked/ emacs package
  - [[file:templates/.travis.yml][.travis.yml]] : The build configuration file for Travis.
+ The =setup-ert= will setup unit test structure:
  - [[file:templates/test/test-helper.el][test-helper]] : Require testing libraries such as =ert=, =el-mock=, helper libraries (=dash= =s=)
    and setup useful variables to different path of the project.
    It also set up a convenience macro =with-sandbox= to perform operation in a test directory.
  - [[file:templates/testpackage-test.el][package-test]] : the file in which the test will have to be written :) (*cpt can't do that for you ;)*)
  - [[file:templates/.ert-runner][.ert-runner]] : configuration of the ert runner. (adding files to the loadpath)
+ The =setup-undercover= will help you set up coverage.

  It will output the code that need to be added to =test/test-helper.el= and =features/support/env.el= so that
  coverage metrics are computed while testing.

# TODO: add ecukes (maybe: call to ecukes new?)
*** Badges
/Cask Package Toolset/ will help you build badges to add to your README, so you can have an idea about
the current state of the package in just a glance.
Calling the =badge= command will output you a list of badges link you can then copy paste to your README.

So far support information such as build status, coverage, melpa status, licence, and also gitter.

/Cask Package Toolset/ supports three different syntaxes: orgmode, markdown and html.
By default you'll get markdown, but you can explicitely specify one of the syntax with the =-s/--syntax= option.

*** Melpa Recipe
If you have not yet published your package to melpa, /Cask Package Toolset/ can easily build up the basic recipe
for you'll just have to add to the [[https://github.com/milkypostman/melpa][melpa repository]].
Note that you might need to add something more than the name and repo (such as extra files to be included)

** Notes

This is no more a prototype, but some features are still added, driven by examples by the *toolsettization* of my own packages.

I will see with the =Cask= team to incorporate it so that that it can be accessed without adding it has Cask dependancy.

*If you have any remark, refactor suggestion or else, just pose an issue ;)*


CHANGELOG.md
Change Log
==========

All notable changes to *cask-package-toolset* will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
*Nothing So Far*

## [0.9.1] - 2016-01-02
### Changed
- Remove `emacs-` prefix when infering project name

## [0.9.0] - 2015-12-20
### Added
- rst syntax support
### Changed
- More dry badge declaration

## [0.8.1] - 2015-12-13
### Changed
- change coverage snippet

## [0.8.0] - 2015-12-11
### Added
- [el-mock](https://github.com/rejeep/el-mock.el) is now setup with unit tests
- FULL COVERAGE :D

## [0.7.0] - 2015-12-10
### Changed
- Update travis build for evm 0.7 + template
### Added
- add command to update travis config

## [0.6.7] - 2015-12-02
### Changed
- Wrong syntax now trigger an error

## [0.6.6] - 2015-11-18
### Fixed
- Typo in repo is now fixed `repositery` -> `repository`

--------------------------------------------------------------------------------

## More detailed history digging would be done one day
### [0.6.5] - 2015-07-08
### [0.6.4] - 2015-07-04
### [0.6.3] - 2015-06-24
### [0.6.2] - 2015-06-23
### [0.6.1] - 2015-06-23
### [0.6] - 2015-06-23
### [0.5.1] - 2015-06-12
### [0.5] - 2015-06-02
### [0.4.2] - 2015-06-02
### [0.4.1] - 2015-05-19
### [0.4] - 2015-05-17
### [0.3.3] - 2015-05-17
### [0.3.2] - 2015-05-16
### [0.3.1] - 2015-05-16
### [0.3] - 2015-05-16
### [0.2.2] - 2015-05-16
### [0.2.1] - 2015-05-15
### [0.2] - 2015-05-15
### [0.1.0] - 2015-05-15

[unreleased]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.9.1...master
[0.9.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.9.0...v0.9.1
[0.9.0]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.8.1...v0.9.0
[0.8.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.8.0...v0.8.1
[0.8.0]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.7...v0.7.0
[0.6.7]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.6...v0.6.7
[0.6.6]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.5...v0.6.6
[0.6.5]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.4...v0.6.5
[0.6.4]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.3...v0.6.4
[0.6.3]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.2...v0.6.3
[0.6.2]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6.1...v0.6.2
[0.6.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.6...v0.6.1
[0.6]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.5.1...v0.6
[0.5.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.5...v0.5.1
[0.5]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.4.2...v0.5
[0.4.2]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.4...v0.4.1
[0.4]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.3.3...v0.4
[0.3.3]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.3...v0.3.1
[0.3]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.2.2...v0.3
[0.2.2]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.2...v0.2.1
[0.2]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/v0.1.0...v0.2
[0.1.0]: https://github.com/AdrieanKhisbe/cask-package-toolset.el/compare/53e803b...v0.1.0


test/test-helper.el
;; test helper inspired from https://github.com/tonini/overseer.el/blob/master/test/test-helper.el

;;; Code:

(require 'f)

(defvar cpt-path
  (f-parent (f-this-file)))

(defvar cpt-test-path
  (f-dirname (f-this-file)))

(defvar cpt-root-path
  (f-parent cpt-test-path))

(defvar cpt-sandbox-path
  (f-expand "sandbox" cpt-test-path))

(when (f-exists? cpt-sandbox-path)
  (error "Something is already in %s. Check and destroy it yourself" cpt-sandbox-path))

(defmacro within-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(let ((default-directory cpt-sandbox-path))
     (when (f-exists? cpt-sandbox-path)
       (f-delete default-directory :force))
     (f-mkdir cpt-sandbox-path)
     ,@body
     (f-delete default-directory :force)))

(defvar commander-ignore t
  "With this prevent commander to run without asking")

(require 'ert)
(require 'el-mock)
(eval-when-compile
  (require 'cl))
(require 's)
(require 'dash)
(require 'undercover)
(undercover "*.el"
            (:exclude "*-test.el")
            (:send-report nil)
            (:report-file "/tmp/undercover-report.json"))
(require 'cask-package-toolset)

(provide 'test-helper)

;;; test-helper.el ends here


templates/test/test-helper.el
;;; test-helper --- Test helper for ${package-name}

;;; Commentary:
;; test helper inspired from https://github.com/tonini/overseer.el/blob/master/test/test-helper.el

;;; Code:

(require 'f)

(defvar cpt-path
  (f-parent (f-this-file)))

(defvar ${package-name}-test-path
  (f-dirname (f-this-file)))

(defvar ${package-name}-root-path
  (f-parent ${package-name}-test-path))

(defvar ${package-name}-sandbox-path
  (f-expand "sandbox" ${package-name}-test-path))

(when (f-exists? ${package-name}-sandbox-path)
  (error "Something is already in %s. Check and destroy it yourself" ${package-name}-sandbox-path))

(defmacro within-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(let ((default-directory ${package-name}-sandbox-path))
     (when (f-exists? ${package-name}-sandbox-path)
       (f-delete default-directory :force))
     (f-mkdir ${package-name}-sandbox-path)
     ,@body
     (f-delete default-directory :force)))

(require 'ert)
(require 'el-mock)
(eval-when-compile
    (require 'cl))
(require '${package-name})

(provide 'test-helper)
;;; test-helper.el ends here


templates/test/package-test.el
;;; Test for `${package-name}'

;;; Commentary:
;; These are the tests for `${package-name}'

;;; Code:

(ert-deftest ${package-name}-should-not-pass ()
  (should-not nil))



features/support/env.el
(require 'f)
(require 'noflet)

(defvar cask-package-toolset-support-path
  (f-dirname load-file-name))

(defvar cask-package-toolset-features-path
  (f-parent cask-package-toolset-support-path))

(defvar cask-package-toolset-root-path
  (f-parent cask-package-toolset-features-path))

(defvar cask-package-toolset-bin-path
  (f-expand "bin" cask-package-toolset-root-path))

(defvar cask-package-toolset-executable
  (f-expand "package-toolset" cask-package-toolset-bin-path))

(defvar cask-package-toolset-sandbox-path
  (f-expand "sandbox" cask-package-toolset-root-path))

;; §todo: set it with a given!
(defvar cask-package-toolset-project-path
  cask-package-toolset-root-path)

(defvar cask-package-toolset-stderr)
(defvar cask-package-toolset-stdout)

(defmacro within-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(let ((default-directory cask-package-toolset-sandbox-path))
     (when (f-exists? cask-package-toolset-sandbox-path)
       (f-delete default-directory :force))
     (f-mkdir cask-package-toolset-sandbox-path)
     ,@body
     (f-delete default-directory :force)))

(add-to-list 'load-path cask-package-toolset-root-path)

(require 'undercover)
(undercover "*.el"
            (:exclude "*-test.el")
            (:report-file "/tmp/undercover-report.json"))

(defvar commander-ignore t
  "With this prevent commander to run without asking")
(require 'cask-package-toolset)
(require 'espuds)
(require 'ert)
(require 'ansi-color)

(Setup
 ;; Before anything has run
 )

(Before
 ;; Before each scenario is run
 )

(After
 ;; After each scenario is run
 )

(Teardown
 ;; After when everything has been run
 )


features/step-definitions/cask-package-toolset-steps.el
;; This file contains your project specific step definitions. All
;; files in this directory whose names end with "-steps.el" will be
;; loaded automatically by Ecukes.

;; cask exec bin/package-toolset
;; §note: was tweak to enable coverage info to be collected!
(When "^I run package-toolset \"\\([^\"]*\\)\"$"
  (lambda (command)
    (let* ((buffer-name "*cpt-output*")
           (buffer
            (progn
              (when (get-buffer buffer-name)
                (kill-buffer buffer-name))
              (get-buffer-create buffer-name)))
           (default-directory (file-name-as-directory cask-package-toolset-project-path))
           (args
            (unless (equal command "")
              (s-split " " command)))
           (commander-ignore nil))
      (noflet ((message (msg &rest args)
                        (with-current-buffer buffer
                          (insert (apply #'format msg args))
                          (newline))))
               (commander-parse args))
      (with-current-buffer buffer
        (let ((content (ansi-color-filter-apply (buffer-string))))
          ;;(cond ((= exit-code 0) ; §later: see how to handle error?
                 (setq cask-package-toolset-stdout content))))))

(When "^I run package-toolset \"\\([^\"]+\\)\" within a new dir$"
      (lambda (command)
        (within-sandbox
         (let ((cask-package-toolset-project-path cask-package-toolset-sandbox-path))
           (When (format "I run package-toolset \"%s\"" command))))))

(Then "^I should see command output:$"
      (lambda (expected)
        (cask-package-toolset-should-match expected cask-package-toolset-stdout)))

(Then "^I should see command error:$"
      (lambda (expected)
        (cask-package-toolset-should-match expected cask-package-toolset-stderr)))

(defun cask-package-toolset-should-match (needle haystack)
  (should (s-contains? needle haystack)))


cask-package-toolset.el
;;; cask-package-toolset.el --- Toolsettize your package   -*- lexical-binding: t; -*-

;; Copyright (C) 2015  Adrien Becchis

;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Created:  2015-05-14
;; Version: 0.9.1
;; Keywords: convenience, tools
;; Url: http://github.com/AdrieanKhisbe/cask-package-toolset.el
;; Package-Requires: ((emacs "24") (cl-lib "0.3") (s "1.6.1") (dash "1.8.0") (f "0.10.0") (commander "0.2.0") (ansi "0.1.0") (shut-up "0.1.0"))

;; This file is not part of GNU Emacs.

;;; Licence:

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Setup Continuous integration for EmacsLisp without pain!
;;
;; Easily setup test and continous intergration to have even better
;; emacs libraries where it's more easy to contribute and build new features! :)

;; Changelog is available at the root of the repos and is named CHANGELOG.md

;;; Code:

(require 'cl-lib)
(require 's)
(require 'dash)
(require 'f)
(require 'commander)
(require 'ansi)
(require 'shut-up)

(when noninteractive
  (shut-up-silence-emacs))

(defvar cask-package-toolset-templates '("Makefile" ".gitignore" ".travis.yml")
  "List of templates supported by `cask-package-toolset'.")
;; §maybe: later replace with alist struct: option, category

(defvar cask-package-toolset-github-remote "origin"
  "Name of the Github remote.")

(defvar cask-package-toolset-github-repository nil
  "Name of the Github repository.")

(defvar cask-package-toolset-badge-syntax :markdown)

(defvar cask-package-toolset-template-dir (f-expand "templates" (f-dirname (f-this-file)))
  "Folder holding the package templates.")

(defvar cask-package-toolset-force nil
  "Forcing the install or not.")

(defconst cask-package-toolset-badge-data-alist
  '(
    (:travis . ((:img . "https://travis-ci.org/${repository-name}.svg")
                (:alt . "Build Status")
                (:link . "https://travis-ci.org/${repository-name}")))
    (:melpa . ((:img . "http://melpa.org/packages/${package-name}-badge.svg")
               (:alt . "MELPA")
               (:link . "http://melpa.org/#/${package-name}")))
    (:melpa-stable . ((:img . "http://stable.melpa.org/packages/${package-name}-badge.svg")
                      (:alt . "MELPA Stable")
                      (:link . "http://stable.melpa.org/#/${package-name}")))
    (:licence . ((:img . "http://img.shields.io/:license-gpl3-blue.svg")
                 (:alt . "License")
                 (:link . "http://www.gnu.org/licenses/gpl-3.0.html")))
    (:gitter . ((:img . "https://badges.gitter.im/Join%20Chat.svg")
                (:alt . "Join the chat")
                (:link . "https://gitter.im/${repository-name}")))
    (:coveralls . ((:img . "https://coveralls.io/repos/${repository-name}/badge.svg")
                   (:alt . "Coverage Status")
                   (:link . "https://coveralls.io/r/${repository-name}")))

    (:github . ((:img . "https://img.shields.io/github/tag/${repository-name}.svg")
                (:alt . "Tag Version")
                (:link . "https://github.com/${repository-name}/tags")))
    )
  "Template string alist for the Badges.")
;; §maybe: replace with custom

(defconst cask-package-toolset-url-syntax-alist
  '((:html . "<a href=\"${link}\"><img alt=\"${alt}\" src=\"${img}\"/></a>")
    (:markdown . "[![${alt}](${img})](${link})")
    (:orgmode . "[[${link}][file:${img}]]")
    (:rst . ".. image:: ${img}\n   :target: ${link}")))


(defun cask-package-toolset-badge-template (name syntax)
  "Return the template for NAME in specified SYNTAX.
Throw exception if non existing!"
  (let* ((syntax-badge-template (cdr (assoc syntax cask-package-toolset-url-syntax-alist)))
         (badge-data (cdr (assoc name cask-package-toolset-badge-data-alist))))
    ;(error "%s" badge-data)
    (s-format syntax-badge-template
              (lambda (var-name)
                (cdr-safe (assoc (intern-soft (s-concat ":" var-name)) badge-data))))))

(defun cask-package-toolset-set-github-repository (repo)
  "Set github REPO."
  ;; §todo: check real repo
  (setq cask-package-toolset-github-repository repo))

(defun cask-package-toolset-set-github-remote (remote)
  "Set github REMOTE."
  ;; §todo: check real remote
  (setq cask-package-toolset-github-remote remote))

(defun cask-package-toolset-set-badge-syntax (syntax)
  "Set badge SYNTAX."
  (if (-any? (lambda(s)(s-equals? s syntax)) '("html" "markdown" "orgmode" "rst"))
      (setq cask-package-toolset-badge-syntax (intern (s-concat ":" syntax)))
      (error (ansi-red (format "%s is not a valid syntax: valid ones [html markdown orgmode rst]" syntax)))))

(defun cask-package-toolset-set-force ()
  "Set force mode."
  (setq cask-package-toolset-force t))

(defun cask-package-toolset-install-template (template-name)
  "Install provided TEMPLATE-NAME."
  (if (cask-package-toolset-template-present-p template-name)
      (progn
        (warn "File %s is already existing. Skipping" template-name)
        ;; §TODO: force option
        nil)
    (progn (cask-package-toolset-copy-template template-name)
           t)))

(defun cask-package-toolset--template-path (template-name)
  "Return path for TEMPLATE-NAME."
  (f-expand template-name cask-package-toolset-template-dir))

(defun cask-package-toolset-template-present-p (template-name)
  "Return t if TEMPLATE-NAME is already present in current-dir."
  (f-exists? (f-expand template-name)))

(defun cask-package-toolset-copy-template (template-name &optional subfolder)
  "Copy specified TEMPLATE-NAME in current folder or specified SUBFOLDER."
  ;; §TODO: add force param? (see how to use a symbol anywhere?)
  (let ((template-file (cask-package-toolset--template-path template-name))
        (destination-file (f-expand template-name)))
    (unless (f-exists? template-file)
      (error "Template %s not found" template-name)) ; §todo: use real error.
    (if (f-exists? destination-file)
        (error "File already existing %s" destination-file)
      (f-copy template-file destination-file))))

(defun cask-package-toolset-install-test-template (package-name)
  "Install the test templates for the PACKAGE-NAME."
  ;; §see: force argument? (for now suppose ok)
  (let ((data `(("package-name" . ,package-name))))
    (f-mkdir "test")
    (f-write-text (cask-package-toolset-fill-template "test/test-helper.el" data)
                  last-coding-system-used
                  "test/test-helper.el")
    (f-write-text (cask-package-toolset-fill-template "test/package-test.el" data)
                  last-coding-system-used
                  (format "test/%s-test.el" package-name))
    (cask-package-toolset-copy-template ".ert-runner")))

(defun cask-package-toolset-fill-template (template-path data)
  "Return filled the template located at TEMPLATE-PATH populated with DATA."
  ;;; §maybe: add argument to retrieve what we need
  (let ((template-file (cask-package-toolset--template-path template-path)))
    (unless (f-exists? template-file)
      (error "Template %s not found" template-path))
    (s-format (f-read-text template-file) 'aget data)))

;; Project Property extractors
;; §todo: maybe: set as a variable
(defun cask-package-toolset-github-repository-name()
  (if cask-package-toolset-github-repository
      cask-package-toolset-github-repository
    (let ((remote-name
           (s-trim
             (shell-command-to-string
              (format "git config --get remote.%s.url"
                      cask-package-toolset-github-remote)))))
      (when (s-contains? "github" remote-name)
        (s-chop-suffix
         ".git"
         (if (s-starts-with? "git" remote-name)
             (nth 1 (s-split ":" remote-name)) ; git protocol
           (nth 1 (s-split ".com/" remote-name)))))))) ; http

(defun cask-package-toolset-github-url (repository-name)
  "Return the github url from REPOSITORY-NAME."
  (unless (s-blank? repository-name)
    (format "http://github.com/%s" repository-name)))

(defun cask-package-toolset-travis-url (repository-name)
  "Return the travis url from REPOSITORY-NAME."
  (unless (s-blank? repository-name)
    (format "http://travis-ci.org/%s" repository-name)))

(defun cask-package-toolset-project-name (repository-name)
  "Return the project name from REPOSITORY-NAME.

Note it remove trailing .el and emacs- suffix if any"
  (unless (s-blank? repository-name)
    (s-chop-prefix "emacs-" (s-chop-suffix ".el" (nth 1 (s-split "/" repository-name))))))

;; Fragment generator
(defun cask-package-toolset-melpa-recipe (repository-name)
  "Return a melpa recipe corresponding to the REPOSITORY-NAME."
  (unless (s-blank? repository-name)
    (format "(%s :fetcher github :repo \"%s\")"
            (cask-package-toolset-project-name repository-name)
            repository-name)))

(defun cask-package-toolset-format-badge (badge repository-name syntax)
  "Return a BADGE corresponding to the REPOSITORY-NAME in specified SYNTAX."
  (unless (s-blank? repository-name)
    (let* ((project-name (cask-package-toolset-project-name repository-name))
           (formated-link (cask-package-toolset-badge-template badge syntax)))
      (s-format formated-link 'aget `(("package-name" . ,project-name)
                                      ("repository-name" . ,repository-name))))))

;; Commands
(defun cask-package-toolset-install-all-templates ()
  "Install all the packages.  (from `cask-package-toolset-templates')."
  (-each cask-package-toolset-templates
    (lambda (template) (cask-package-toolset-install-template template))))

(defun cask-package-toolset-noop ()
  "Invite to specify a command."
  (message (ansi-blue "Give us a command. install for instance, or consult usage with help")))

(defun cask-package-toolset-print-github-remote()
  (message "%s" (cask-package-toolset-github-repository-name)))

(defun cask-package-toolset-print-melpa-recipe ()
  "Print Melpa Recipe."
  (let ((melpa-recipe (cask-package-toolset-melpa-recipe
                       (cask-package-toolset-github-repository-name))))
    (message (if (s-blank? melpa-recipe)
                 (ansi-red "We could not retrieve melpa recipe, specify the remote if origin does not refer to your github repository.")
               melpa-recipe))))

(defun cask-package-toolset-setup-test()
  "Install Ert tests if not yet existing"
  (let ((package-name (cask-package-toolset-project-name
                       (cask-package-toolset-github-repository-name))))
    (if (s-blank? package-name)
       (message (ansi-red "We could not retrieve project-name from github repo, specify the remote if origin does not refer to your github repository."))
        (if (or (not (f-exists? (f-expand "test")))
                cask-package-toolset-force)
        (progn
            (cask-package-toolset-install-test-template package-name)
              (message (ansi-green "Ert Scaffold files generated")))
          (message (ansi-red "Some test file already exist. If you want to erase them, add --force option"))))))

(defun cask-package-toolset-print-setup-coverage()
  "Install Ert tests if not yet existing"
  (let ((package-name (cask-package-toolset-project-name
                       (cask-package-toolset-github-repository-name))))
    (if (s-blank? package-name)
        (message (ansi-red "We could not retrieve project-name from github repo, specify the remote if origin does not refer to your github repository."))
      (message "%s" (cask-package-toolset-fill-template "undercover.el"
                                                        `(("package-name" . ,package-name)))))))

(defun cask-package-toolset-ensure-latest-travis-config()
  "Updare travis recipe if up to date"
  (if (f-exists? ".travis.yml")
      (let ((travis-content (f-read-text ".travis.yml")))
        (if (s-contains? "/ebcd57c3af83b049833b/" travis-content)
            (message (ansi-green "Travis config already updated"))
          (let ((template-file (cask-package-toolset--template-path ".travis.yml"))
                (destination-file (f-expand ".travis.yml")))
            (message "Updating travis config\nYou might need to check the emacs version you want to support")
            (f-delete destination-file)
            (f-copy template-file destination-file))))
    (error "Travis not setup yet, run `setup' instead")))

(defun cask-package-toolset-print-badges ()
  "Print Melpa Recipe."
  (let ((repository-name (cask-package-toolset-github-repository-name)))
    (if (not (s-blank? repository-name))
        (-each (-map (lambda (badge-type)
                       (cask-package-toolset-format-badge badge-type repository-name cask-package-toolset-badge-syntax))
                     '(:travis :coveralls :melpa :melpa-stable :github :licence :gitter))
          (lambda (badge) (message "%s" badge)))
      (message (ansi-red "We could not retrieve melpa recipe, specify the remote if origin does not refer to your github repository.")))))

(defun cask-package-toolset-print-status()
  "Print current status of the repository"
  (let ((file-status
         `(;; File/Name - Status - Reco if Nok
           ;; check installed templates.
           ,@(-map (lambda (file)(list file (f-exists? file) "You should run `package-toolset setup'"))
                   cask-package-toolset-templates)
           ;; check installed ert
           ("Ert Test" ,(f-directory? "test") "You should run `cask exec package-toolset setup-ert'")
           ;; check installed ecukes
           ("Ecukes Features" ,(and (f-directory? "features/step-definitions")
                                    (f-directory? "features/support"))
            "You should run `cask exec ecukes new'")
           )))
    (-each file-status
      (lambda (it)
        (message "- %s →   %s" (s-pad-right 28 " " (ansi-blue (car it))) (if (nth 1 it) (ansi-green "Ok") (ansi-red (nth 2 it))))))))

(commander
 (name "cask-package-toolset")
 (description "Toolsettize your emacs package")
 (config ".cask-package-toolset")
 (default cask-package-toolset-noop)

 (option "--github-repo <repo>, -g <repo>" cask-package-toolset-set-github-repository)
 (option "--remote <remote>, -r <remote>" cask-package-toolset-set-github-remote)
 (option "--syntax <syntax>, -s <syntax>" cask-package-toolset-set-badge-syntax)
 (option "--force, -f" cask-package-toolset-set-force)
 (option "--help, -h" commander-print-usage) ; §todo: option specific help

 (command "status" "CI status of your package" cask-package-toolset-print-status)
 (command "install" "Install basic template for CI" cask-package-toolset-install-all-templates)
 (command "setup" "Install basic template for CI" cask-package-toolset-install-all-templates)
 (command "setup-ert" "Setup ert test structure" cask-package-toolset-setup-test)
 (command "setup-undercover" "Print snippets of code to set up coverage" cask-package-toolset-print-setup-coverage)
 (command "badge" "Print badges to add to your READLE" cask-package-toolset-print-badges)
 (command "melpa-recipe" "Print recipe for melpa" cask-package-toolset-print-melpa-recipe)
 (command "git" "Print deduced github remote" cask-package-toolset-print-github-remote)
 (command "update-travis" "Update travis recipe if needed" cask-package-toolset-ensure-latest-travis-config)
 (command "help" "Show usage information" commander-print-usage))


(provide 'cask-package-toolset)
;;; cask-package-toolset.el ends here
